Add Comprehensive Developer Guides and Best Practices

ypszn opened the pull request:
## Summary
Added three comprehensive developer guides to improve the Fluent development experience:

- **Common Patterns & Best Practices**: Covers error handling, gas optimization, security, testing, and debugging
- **Troubleshooting Guide**: Solutions for common build, deployment, and runtime issues  
- **Complete Blended App Tutorial**: Step-by-step tutorial for building hybrid Rust/Solidity applications

## Why This is Needed
The existing developer guides were basic and lacked comprehensive examples. These new guides provide:
- Real-world patterns and best practices
- Solutions to common developer problems
- Complete tutorial examples that developers can follow

## Testing
✅ All guides tested locally and render correctly
✅ Build passes without errors (`npm run build` successful)
✅ Navigation and links work properly
✅ Code examples are properly formatted

## Impact
Significantly improves developer onboarding and reduces friction when building on Fluent.

## Files Added
- `docs/developer-guides/common-patterns.md`
- `docs/developer-guides/troubleshooting-guide.md` 
- `docs/developer-guides/complete-blended-app-tutorial.md`

ypszn committed:
feat: add comprehensive developer guides and best practices

tmsdkeys commented:
gblend @ypszn and thank you for your contribution, I will have to take a closer look at this when I have some bandwidth. will keep you posted

ypszn commented:
gblend @tmsdkeys. okay, no problem

tmsdkeys commented on docs/developer-guides/common-patterns.md:
````diff
@@ -0,0 +1,691 @@
+}
+```
+
+### Solidity Error Handling
````
Not sure if I'd want to add this here. There are better resources out there to learn about Solidity best practices

tmsdkeys commented on docs/developer-guides/common-patterns.md:
```diff
@@ -0,0 +1,691 @@
+
+### Rust Contract Optimization
+
+#### 1. Efficient Storage Patterns
```
These are not the canonical way to manage storage with the Fluentbase SDK, and there's already best practices in https://docs.fluent.xyz/fluentbase-sdk/storage#generated-api

tmsdkeys commented on docs/developer-guides/common-patterns.md:
````diff
@@ -0,0 +1,691 @@
+}
+```
+
+### Solidity Gas Optimization
````
Ditto regarding Solidity best practices and patterns

tmsdkeys commented on docs/developer-guides/common-patterns.md:
````diff
@@ -0,0 +1,691 @@
+}
+```
+
+## Testing Strategies
````
The need for testing docs to be added is valid, yet this is not the correct way to do it

tmsdkeys commented on docs/developer-guides/troubleshooting-guide.md:
````diff
@@ -0,0 +1,570 @@
+
+```toml
+[dependencies]
+fluentbase-sdk = "0.3.6"  # Use the latest compatible version
````
at 0.4.3-dev currently

tmsdkeys commented on docs/developer-guides/troubleshooting-guide.md:
````diff
@@ -0,0 +1,570 @@
+]
+```
+
+### gblend Build Issues
````
Would refer here to https://docs.fluent.xyz/gblend/troubleshooting to avoid duplication

tmsdkeys commented on docs/developer-guides/troubleshooting-guide.md:
```diff
@@ -0,0 +1,570 @@
+gblend estimate-gas --contract YourContract
+
+# Deploy with higher gas limit
+gblend deploy --gas-limit 5000000
```
The deploy command is `gblend create`

tmsdkeys commented on docs/developer-guides/troubleshooting-guide.md:
```diff
@@ -0,0 +1,570 @@
+- Optimize storage layout
+- Avoid unbounded loops
+
+#### 2. Network Configuration Issues
```
Would opt for a different solution here, e.g.

gblend <command> \
--rpc-url $RPC_URL

tmsdkeys commented on docs/developer-guides/troubleshooting-guide.md:
````diff
@@ -0,0 +1,570 @@
+gblend config --rpc-url https://rpc.dev.gblend.xyz
+```
+
+#### 3. Contract Verification Failures
````
These are not the correct commands, you could refer to https://docs.fluent.xyz/gblend/usage#verification

tmsdkeys commented on docs/developer-guides/troubleshooting-guide.md:
```diff
@@ -0,0 +1,570 @@
+
+**Problem**: Solidity interface not generated correctly.
+
+**Solution**: Check your Rust contract structure:
```
This is not very specific

tmsdkeys commented on docs/developer-guides/troubleshooting-guide.md:
```diff
@@ -0,0 +1,570 @@
+- `bool` → `bool`
+- `string` → `String`
+
+### Storage Access Issues
```
Again, storage with Fluentbase SDK is handled differently

tmsdkeys commented on docs/developer-guides/troubleshooting-guide.md:
```diff
@@ -0,0 +1,570 @@
+- `"functionName(address,uint256)"` - Multiple parameters
+- `"functionName(uint256[])"` - Array parameter
+
+### 4. Storage Initialization
```
Ditto regarding storage

tmsdkeys commented on docs/developer-guides/troubleshooting-guide.md:
```diff
@@ -0,0 +1,570 @@
+
+### High Gas Consumption
+
+#### 1. Inefficient Storage Operations
```
Ditto regarding storage

tmsdkeys commented on docs/developer-guides/complete-blended-app-tutorial.md:
```diff
@@ -0,0 +1,384 @@
+sidebar_position: 6
+---
+
+Complete Blended App Tutorial
```
This developer guide is similar but less detailed than something we're internally working on, so I'm leaning towards not including this guide here. 

What I would suggest here is that you add this guide as a README to the project repo (fully functioning) and then you can create a PR to https://github.com/fluentlabs-xyz/awesome-fluent to have it added as a community project

tmsdkeys reviewed (changes requested):
I've made a first round of reviews here. A couple of high-level notes:

- I'll add this more specifically to the contribution guidelines, but for features additions like this it would have been good to add an issue first to discuss scope of the addition to avoid redundant work
- Can I ask if this was generated with the help of AI tools? Seems like some of the code here was hallucinated... I think it's fine to leverage AI tools (we do internally) but it would be good to: 1) make a mention of this and 2) review everything personally before submitting

In general, I am in favour of adding common-patterns and troubleshooting sections but arguably the common patterns would be better situated in the Fluentbase SDK section (I would refer to external sources for Solidity best practices and common patterns). The troubleshooting guide does make sense to me as a guide and is a good extension of the gblend troubleshooting section.

As mentioned in a comment, the guide with the Hybrid Token Exchange is something we're working on in the Fluent workshop series https://github.com/tmsdkeys/fluent-workshop-series and I am inclined to use that as canonical version, however I suggested adding your project as a community contribution in awesome-fluent.

In conclusion, for next steps:
- Please make requested changes + testing to common-patterns and we can see if it makes sense as a guide or move it around
- Please make revision to troubleshooting-guide.md
- Please migrate the token exchange guide to its own project repo and open a PR in awesome-fluent.

Thanks again for your contribution!

ypszn committed:
fix: update storage patterns & reorganize developer guides

ypszn commented:
Hello @tmsdkeys  I apologize for not mentioning it but i did use AI tools to assist me with some of the codes and i've done a proper review before pushing these new changes. 

## Changes Made
- Fixed storage access issues and how storage is handled with Fluentbase SDK
- Updated troubleshooting guide with the correct Fluentbase SDK patterns  
- Moved `common-patterns` to Fluentbase SDK section
- Removed duplicate guides and improved structure

I made changes to everything you pointed out and i've removed the hybrid token exchange section (working on creating a repo for it and open a PR in awesome-fluent)

tmsdkeys commented:
Thanks for applying changes @ypszn , had a bit of a backlog but should be able to review again in the upcoming days!
